### DISCUSSION TITLE
Spacy version 3 model training

### DISCUSSION THREAD

**Discussion by mandar-avhad** [explosion/spaCy#8764] [Help: Coding & Implementations] [answered]
How to train spacy model for text classification in version 3 without the CLI method?
I'm trying to train the model in jupyter notebook.
Please share the method and code if any.

**Comment by polm** [comment, marked as answer]
There's a sample training loop in the docs, and you can look at how training works in spaCy internally if you want to write your own loop. However we recommend against this because it's easy to get things wrong. 

Can you clarify why you don't want to use the command line method? You can run command line commands in Colab or Jupyter notebooks by just prefixing them with `!`.

**Reply by mandar-avhad** [reply]
Can you please share the link of the training loop doc?

**Reply by polm** [reply]
In the training models section of the docs, at the bottom there's a section on [the internal training API](https://spacy.io/usage/training#api) with an example training loop.

**Reply by polm** [reply]
Also, if you tried something and got errors like you posted in the other thread, please provide the code that generated the errors if you want help. Without the code to reference there's no real way to tell you what to fix.

**Reply by mandar-avhad** [reply]
Code:
 
```
from spacy.util import minibatch, compounding

##Disabling other components
#other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'textcat']
#with nlp.disable_pipes(*other_pipes):  # only train textcat

##only selecting textcat for training
with nlp.select_pipes(enable="textcat"):
    #optimizer = nlp.initialize(lambda: examples)
    optimizer = nlp.resume_training()
    print("Training the model...")
    print('{:^5}\t{:^5}\t{:^5}\t{:^5}'.format('LOSS', 'P', 'R', 'F'))

    # Performing training
    for i in range(n_iter):
        losses = {}
        batches = minibatch(examples, size=compounding(4.0, 16.0, 1.001))
        for batch in batches:
            nlp.update(batch, drop=0.4, losses=losses)
        #print(losses)
        # Calling the evaluate() function and printing the scores
        #with textcat.model.use_params(optimizer.averages):
        scores = evaluate(nlp.tokenizer, textcat, dev_texts, dev_cats)
        print('{0:.3f}\t{1:.3f}\t{2:.3f}\t{3:.3f}'.format(losses['textcat'], scores['textcat_p'], scores['textcat_r'], scores['textcat_f']))
```

Errors:
When trying to load the spacy pretrained en_core model and train it on new data, getting the either of these errors.
1. optimizer = nlp.resume_training()
ValueError: Cannot get dimension 'nO' for model 'sparse_linear': value unset

2. optimizer = nlp.initialize(lambda: examples)
ValueError: [E955] Can't find table(s) lexeme_norm for language 'en' in spacy-lookups-data. Make sure you have the package installed or provide your own lookup tables if no default lookups are available for your language.

**Reply by polm** [reply]
ValueError: Cannot get dimension 'nO' for model 'sparse_linear': value unset

Sounds like you need to call initialize on the components. If you use the config / cli training this is taken care of for you.

    ValueError: [E955] Can't find table(s) lexeme_norm for language 'en' in spacy-lookups-data. Make sure you have the package installed or provide your own lookup tables if no default lookups are available for your language.

Did you check if you have spacy-lookups-data installed?

**Reply by mandar-avhad** [reply]
Yes, it worked.
Thank you :)

**Reply by mandar-avhad** [reply]
Will verify again
